### PR TITLE
fix(executor): Remove IsTransientErr check for ExponentialBackoff. Fixes #4144

### LINKED ISSUE #4144: Add request timed out in errors.isTransientNetworkErr
## Summary 

What happened/what you expected to happen?

We got following errors from time to time:
`
time="2020-09-25T14:34:34.038Z" level=error msg="executor error: etcdserver: request timed out"
`

Looking at the ExponentialBackoff usage in executor codes, e.g. https://github.com/argoproj/argo/blob/master/workflow/executor/executor.go#L622

It will failed earlier if the error is not transient error, the method `IsTransientErr`: https://github.com/argoproj/argo/blob/master/util/errors/errors.go#L40
Seems missing case for `request timed out`

## Diagnostics

What version of Argo Workflows are you running?

```yaml
2.10.1
```

```
time="2020-09-25T14:06:14.611Z" level=info msg="Waiting on main container"
time="2020-09-25T14:06:15.541Z" level=info msg="main container started with container ID: b69687396a67edffa59337f84882f22ded284b44b05b61718578bd0c403f207e"
time="2020-09-25T14:06:15.541Z" level=info msg="Starting annotations monitor"
time="2020-09-25T14:06:15.544Z" level=info msg="Waiting for container b69687396a67edffa59337f84882f22ded284b44b05b61718578bd0c403f207e to complete"
time="2020-09-25T14:06:15.544Z" level=info msg="Starting to wait completion of containerID b69687396a67edffa59337f84882f22ded284b44b05b61718578bd0c403f207e ..."
time="2020-09-25T14:06:15.544Z" level=info msg="Starting deadline monitor"
time="2020-09-25T14:11:14.612Z" level=info msg="Alloc=6049 TotalAlloc=42920 Sys=71872 NumGC=12 Goroutines=10"
time="2020-09-25T14:16:14.611Z" level=info msg="Alloc=8286 TotalAlloc=70981 Sys=72384 NumGC=18 Goroutines=10"
time="2020-09-25T14:21:14.612Z" level=info msg="Alloc=6010 TotalAlloc=99087 Sys=72384 NumGC=25 Goroutines=10"
time="2020-09-25T14:26:14.611Z" level=info msg="Alloc=7707 TotalAlloc=127181 Sys=72640 NumGC=31 Goroutines=10"
time="2020-09-25T14:31:14.612Z" level=info msg="Alloc=5406 TotalAlloc=155271 Sys=72640 NumGC=38 Goroutines=10"
time="2020-09-25T14:34:34.038Z" level=warning msg="Failed to wait for container id 'b69687396a67edffa59337f84882f22ded284b44b05b61718578bd0c403f207e': etcdserver: request timed out"
time="2020-09-25T14:34:34.038Z" level=error msg="executor error: etcdserver: request timed out"
time="2020-09-25T14:34:34.038Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2020-09-25T14:34:34.038Z" level=info msg="Capturing script exit code"
time="2020-09-25T14:34:34.038Z" level=info msg="Getting exit code of b69687396a67edffa59337f84882f22ded284b44b05b61718578bd0c403f207e"
time="2020-09-25T14:34:34.038Z" level=info msg="Annotations monitor stopped"
time="2020-09-25T14:34:34.044Z" level=info msg="No output parameters"
time="2020-09-25T14:34:34.044Z" level=info msg="No output artifacts"
time="2020-09-25T14:34:34.044Z" level=info msg="Killing sidecars"
time="2020-09-25T14:34:34.105Z" level=info msg="Alloc=6181 TotalAlloc=173280 Sys=72640 NumGC=42 Goroutines=9"

```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by anggao**
fix(executor): Remove IsTransientErr check for ExponentialBackoff https://github.com/argoproj/argo/issues/4144

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -2936,9 +2936,6 @@ func (woc *wfOperationCtx) deletePDBResource() error {
 		err := woc.controller.kubeclientset.PolicyV1beta1().PodDisruptionBudgets(woc.wf.Namespace).Delete(woc.wf.Name, &metav1.DeleteOptions{})
 		if err != nil && !apierr.IsNotFound(err) {
 			woc.log.WithField("err", err).Warn("Failed to delete PDB.")
-			if !errorsutil.IsTransientErr(err) {
```
Comment (alexec) [inline workflow/controller/operator.go:2939]: please do not change this file - this could create a thundering herd problem

File: util/util.go [hunk 2]
```diff
@@ -40,9 +39,6 @@ func GetSecrets(clientSet kubernetes.Interface, namespace, name, key string) ([]
 		secret, err = secretsIf.Get(name, metav1.GetOptions{})
 		if err != nil {
 			log.Warnf("Failed to get secret '%s': %v", name, err)
-			if !errorsutil.IsTransientErr(err) {
```
Comment (alexec) [inline util/util.go:43]: same here, might be used by code that does not want to retry these errors

File: workflow/controller/operator.go (hunk 1 above)
Comment (anggao) [reply workflow/controller/operator.go]: ah sure, changes reverted.

File: util/util.go (hunk 2 above)
Comment (anggao) [reply util/util.go]: reverted.

**Comment by anggao** [comment]
I am wondering shall we tweak this config

```
var ExecutorRetry = wait.Backoff{
	Steps:    8,
	Duration: 1 * time.Second,
	Factor:   1.0,
	Jitter:   0.1,
}
```
e.g. to make factor > 1.0 ? otherwise the retry will have the same timeout ?

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
Re-running failed jobs.

**Comment by alexec** [comment]
> I am wondering shall we tweak this config
> 
> ```
> var ExecutorRetry = wait.Backoff{
> 	Steps:    8,
> 	Duration: 1 * time.Second,
> 	Factor:   1.0,
> 	Jitter:   0.1,
> }
> ```
> 
> e.g. to make factor > 1.0 ? otherwise the retry will have the same timeout ?



> I am wondering shall we tweak this config
> 
> ```
> var ExecutorRetry = wait.Backoff{
> 	Steps:    8,
> 	Duration: 1 * time.Second,
> 	Factor:   1.0,
> 	Jitter:   0.1,
> }
> ```
> 
> e.g. to make factor > 1.0 ? otherwise the retry will have the same timeout ?

Good idea. 

http://backoffcalculator.com/?attempts=8&rate=1


Retry | Seconds | Timestamp
-- | -- | --
1 | 1.00 | 2020-09-29 12:53:50
2 | 2.00 | 2020-09-29 12:53:51
3 | 3.00 | 2020-09-29 12:53:52
4 | 4.00 | 2020-09-29 12:53:53
5 | 5.00 | 2020-09-29 12:53:54
6 | 6.00 | 2020-09-29 12:53:55
7 | 7.00 | 2020-09-29 12:53:56
8 | 8.00 | 2020-09-29 12:53:57

Maybe?


Retry | Seconds | Timestamp
-- | -- | --
1 | 0.50 | 2020-09-29 12:53:49
2 | 1.50 | 2020-09-29 12:53:50
3 | 3.50 | 2020-09-29 12:53:52
4 | 7.50 | 2020-09-29 12:53:56
5 | 15.50 | 2020-09-29 12:54:04



@jessesuen ?

**Comment by anggao** [comment]
According to https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Backoff

```
	// The initial duration.
	Duration time.Duration
	// Duration is multiplied by factor each iteration, if factor is not zero
	// and the limits imposed by Steps and Cap have not been reached.
	// Should not be negative.
	// The jitter does not contribute to the updates to the duration parameter.
	Factor float64
	// The sleep at each iteration is the duration plus an additional
	// amount chosen uniformly at random from the interval between
	// zero and `jitter*duration`.
	Jitter float64
```
It seems with `Factor:   1.0,` the duration will always around 1 second ?

**Comment by alexec** [comment]
I think your read is correct

**Comment by anggao** [comment]
@alexec what about this config, it will retry 5 times, with maximum wait around 10s

```
var ExecutorRetry = wait.Backoff{
	Steps:    5,
	Duration: 1 * time.Second,
	Factor:   1.6,
	Jitter:   0.5,
}
```

**Comment by alexec** [comment]
I get a different plan?

http://backoffcalculator.com/?attempts=5&interval=1&rate=1.6

Retry | Seconds | Timestamp
-- | -- | --
1 | 1.00 | 2020-10-02 10:43:50
2 | 2.60 | 2020-10-02 10:43:52
3 | 5.16 | 2020-10-02 10:43:54
4 | 9.26 | 2020-10-02 10:43:58
5 | 15.81 | 2020-10-02 10:44:05

**Comment by anggao** [comment]
I feel that UI page maybe not implement the same way as the go module ?
e.g. I thought it should be 1 * 1.6 ^ 5 = 10.48576

**Comment by alexec** [comment]
> I feel that UI page maybe not implement the same way as the go module ?
> e.g. I thought it should be 1 * 1.6 ^ 5 = 10.48576

I think your. Correct, I just used anther one and got: http://exponentialbackoffcalculator.com/


un | Seconds | Timestamp
-- | -- | --
0 | 0.000 | 02/10/2020, 18:17:34:173 UTC
1 | 1.000 | 02/10/2020, 18:17:35:173 UTC
2 | 2.600 | 02/10/2020, 18:17:36:773 UTC
3 | 5.160 | 02/10/2020, 18:17:39:333 UTC
4 | 9.256 | 02/10/2020, 18:17:43:429 UTC

**Comment by alexec** [comment]
@nggao to move this PR forward, do the backoff part in another PR?

**Review by alexec — APPROVED** [review]

**Comment by anggao** [comment]
@alexec Yeah make sense let's do it in another PR

**Comment by anggao** [comment]
https://github.com/argoproj/argo/pull/4196

**Comment by alexec** [comment]
Can you sync with master to force a rebuild?

**Comment by anggao** [comment]
Done, did a rebase

**Comment by anggao** [comment]
@alexec Do you know when this fix will be released, it seems not included in v2.11.5 ?

**Comment by alexec** [comment]
I've cherry-picked to v2.11.6.

**Comment by anggao** [comment]
@alexec  thanks!

**Pull request merged**